streamalert/classifier/clients/firehose.py

```diff
@@ -325,32 +314,29 @@ def generate_firehose_suffix(cls, use_prefix, prefix, log_stream_name):
         Returns:
             str: suffix of stream name
         """
+        if prefix:
+            prefix += '_'
 
-        reserved_len = cls.FIREHOSE_NAME_MAX_LEN
-
-        if use_prefix:
-            # the prefix will have a trailing '_' (underscore) that's why deduct 1
-            # in the end. Please refer to terraform module for more detail
-            # terraform/modules/tf_kinesis_firehose_delivery_stream/main.tf
-            reserved_len = reserved_len - len(prefix) - 1
+        # This same substitution method is used when naming the Delivery Streams
+        stream_name = cls.sanitized_value(cls.DEFAULT_FIREHOSE_FMT.format(prefix, log_stream_name))
+        if len(stream_name) <= cls.AWS_FIREHOSE_NAME_MAX_LEN:
+            return stream_name
 
-        # Don't change the stream name if its length is complaint
-        if len(log_stream_name) <= reserved_len:
-            return log_stream_name
-
-        # Otherwise keep the first 51 chars if no prefix and hash the rest string into 8 chars.
-        # With prefix enabled, keep the first (51 - len(prefix) and hash the rest string into 8
-        # chars.
-        pos = reserved_len - cls.FIREHOSE_NAME_HASH_LEN
-        hash_part = log_stream_name[pos:]
-        hash_result = hashlib.md5(hash_part.encode()).hexdigest() # nosec
+        base_name = stream_name[:cls.AWS_FIREHOSE_NAME_MAX_LEN - cls.FIREHOSE_NAME_MIN_HASH_LEN]
+        if not base_name.endswith('_'):
+            # make sure this ends in an underscore, but not 2
+            base_name = '{}_'.format(
+                base_name[:-1]
+            ) if base_name[-2] != '_' else '{}_'.format(base_name[:-2])
 
         # combine the first part and first 8 chars of hash result together as new
         # stream name.
         # e.g. if use prefix
         # 'very_very_very_long_log_stream_name_abcd_59_characters_long' may hash to
         # 'very_very_very_long_log_stream_name_abcd_59_06ceefaa'
-        return ''.join([log_stream_name[:pos], hash_result[:cls.FIREHOSE_NAME_HASH_LEN]])
+        return '{}{}'.format(
+            base_name, hashlib.md5(base_name.encode()).hexdigest()
+        )[:cls.AWS_FIREHOSE_NAME_MAX_LEN]  # nosec
```

Comment:
Need to move `#nosec` to line 338. CI complaint about use of `md5`.

Reply:
done